CORE-5940: (ledger-common) Encode CPK metadata in transaction metadata

omfR3 opened the pull request:
When persisting transactions, we'll need to store information about the CPI and CPKs used to generate them, to facilitate backchain resolution later. This PR adds some basic information to the transaction:

* CPI info
  * name
  * version
  * file checksum
  * signer summary hash
* Info for each CPK marked as contract
  * name
  * version
  * file checksum
  * signer summary hash

Note -- at the moment we don't have a reliable method of getting the exact version of the CPI that should be associated with a transaction, so we're inserting some dummy data for now. We need to revisit later and fix that.

vlajos committed:
CORE-5091 Consensual Signed Tx signature tweaks

omfR3 committed:
CORE-5940: (Consensual ledger) Add mocks for recently-added services used by ConsensualLedgerServiceImpl and ConsensualLedgerTransactionImpl. Tests still failing.

omfR3 committed:
CORE-5940: (Consensual ledger) Add helper for mocks relating to transactions and tx builder + temporarily roll back some of the tx builder signing changes

omfR3 committed:
CORE-5940: (ledger-common) Add some CPK information (name and checksum) to transaction metadata

omfR3 committed:
CORE-5940: (ledger-common/consensual) Add missing mocks to the other consensual ledger TX tests + handle null CPK list in metadata

omfR3 commented on libs/ledger/ledger-consensual-impl/src/test/kotlin/net/corda/ledger/consensual/impl/ConsensualTransactionMocks.kt:
```diff
@@ -0,0 +1,114 @@
+import java.time.Instant
+import java.util.UUID
+
+class ConsensualTransactionMocks {
```
Didn't expect to have to do so much mocking here... I feel like it introduces too much coupling with these classes, but not sure of the alternative (other than to dump the unit tests and rely on integration tests which might not be so bad).

mnesbit commented on libs/ledger/ledger-consensual-impl/src/main/kotlin/net/corda/ledger/consensual/impl/transaction/ConsensualTransactionBuilderImpl.kt:
```diff
@@ -62,12 +77,27 @@ class ConsensualTransactionBuilderImpl(
-                // TODO(CORE-5940 set CPK identifier/etc)
+                DIGEST_SETTINGS_KEY to WireTransactionDigestSettings.defaultValues,
+                PLATFORM_VERSION_KEY to memberLookup.myInfo().platformVersion,
+                CPK_IDENTIFIERS_KEY to getCpks().map { "${it.cpkId.name}:${it.fileChecksum.toHexString()}" }
```
We need the full CPK: name, version, signerSummaryhash and fileChecksum to be able to do evolution checking

omfR3 replied:
K, I've added those (although signerSummaryHash is nullable). Packing them together into a string like this maybe not the best though. I think we have another task on the board for doing some JSON encoding of the metadata, though.

vlajos replied:
Could you please use(+update) the CpkIdentifier class.
And please save the three CPI related metadata to the TxMetadata too. 
https://github.com/corda/corda-runtime-os/pull/2248/files#diff-f6522e2cf23dedc2a84238461155e24b3a0c76d1f2c196862e9913030778ab82R131-R133

omfR3 replied:
@vlajos good idea. For some reason I assumed we were sticking with atomic values or maps/lists, but using data classes for CPK and CPI metadata makes more sense. I renamed `CpkIdentifier` to `CpkMetadata` and introduced `CpiIdentifier` with similar fields + a `List<CpkMetadata>`, but I'm running into a very odd error:

```
expected CpiMetadata but found class java.util.LinkedHashMap
```
i.e. I'm storing a `CpiMetadata` in the metadata map under the `CPI_METADATA_KEY` (renamed from `CPK_METADATA_KEY` since we have 1 CPI and multiple CPKs), but it's somehow being replaced with another map. I've obviously mixed something silly up somewhere or forgot to update something...

omfR3 replied:
Okay, I think it's a problem with the JSON serialisation when creating the `WireTransaction`. So I guess that's why we were using those weakly-typed maps. Will have to fix this tomorrow -- there's probably overlap with the [JSON serialisation improvements task](https://r3-cev.atlassian.net/browse/CORE-6890) in the backlog.

omfR3 replied:
I've renamed those two classes to `CpiSummary` and `CpkSummary` to avoid collisions.

mnesbit commented on libs/ledger/ledger-consensual-impl/src/main/kotlin/net/corda/ledger/consensual/impl/transaction/ConsensualTransactionBuilderImpl.kt:
```diff
@@ -62,12 +77,27 @@ class ConsensualTransactionBuilderImpl(
-                DIGEST_SETTINGS_KEY to WireTransactionDigestSettings.defaultValues
-                // TODO(CORE-5940 set CPK identifier/etc)
+                DIGEST_SETTINGS_KEY to WireTransactionDigestSettings.defaultValues,
+                PLATFORM_VERSION_KEY to memberLookup.myInfo().platformVersion,
```
This feels wrong. We should be able to get platform version without going through membership info, which might be slightly out of date. I am hoping that we can get an OSGi service that gives us this info in a more direct form.

vlajos replied:
I think this code comes from my signing improvements code. (I have found this approach to get the platformVersion somewhere.)
What is the proper way to get the platform version?

omfR3 replied:
Had a quick discussion on Slack with @mnesbit who pointed out this we don't have a better solution yet, and need to do more design thinking around platform versioning. So I guess we'll stick with this for now, but make a note that we need to revisit at some point.

vlajos replied:
Thanks!

omfR3 commented:
Just noticed the integration tests fail with this:
```
TEST net.corda.ledger.consensual.impl.test.ConsensualLedgerServiceTest <<< ERROR: 1/1 services (objectClass=net.corda.v5.ledger.consensual.ConsensualLedgerService) didn't arrive within 1000ms                   
java.lang.AssertionError: 1/1 services (objectClass=net.corda.v5.ledger.consensual.ConsensualLedgerService) didn't arrive within 1000ms   
```
Which is odd. Last time I saw that, it was due to a dependency being missing from `test.bndrun`, but that doesn't look likely here. Maybe some of the newly-added service dependencies needs to be added to `test.bndrun`, but I'm curious as to why it would mention `ConsensualLedgerService` with no clues beyond that. Debugging OSGi issues can be like searching for the right piece of paper while wearing a blindfold...

### Update
Found a solution in @vlajos's [branch](https://github.com/corda/corda-runtime-os/commit/3ee07f28fa9cab221fb440c152fa6df6a990b4c4), yoink! Just needed to add something to `build.gradle`. Thanks :smile:

omfR3 committed:
CORE-5940: (Consensual ledger) Fix integration test OSGi error

omfR3 committed:
CORE-5940: (Consensual ledger) Add more CPK information to transaction metadata

omfR3 committed:
CORE-5940: Fix detekt errors

omfR3 committed:
CORE-5940: (ledger-common / consensual) Provide some structure to CPI/CPK metadata in transactions rather than mushing it all into a string.

omfR3 committed:
CORE-5940: (ledger-common + consensual) Rename transaction.CpkIdentifier -> transaction.CpkMetadata and use them to encapsulate tx CPK metadata rather than fumbling with ordered maps. Something is still broken.

vlajos committed:
CORE-5091 Instead of using cpi-info-read-service and virtual-node-info-read-service we extract the CPK info directly from sandboxGroupContext.sandboxGroup.metadata

(cherry picked from commit 5929679c5cc81a028b2e6a5806bf1a26ef096b77)

omfR3 committed:
CORE-5940: (ledger-common) Reconstitute CpiMetadata after JSON mapper turns it into a generic Map<String, Any?>

omfR3 committed:
CORE-5940: (ledger-common) Refactor CpiSummary + TransactionMetadata

* Hoist list of CPK metadata (CpkSummary) out of CpiSummary
* Add separate CPK_METADATA_KEY + accessor to TransactionMetaData

omfR3 committed:
CORE-5940: (ledger-consensual) Use accessor to get CPK metadata so we can persist it later

vlajos commented on libs/ledger/ledger-common-impl/src/main/kotlin/net/corda/ledger/common/impl/transaction/CpiSummary.kt:
```diff
@@ -0,0 +1,11 @@
+    val version: String,
+    val signerSummaryHash: String?,
+    val fileChecksum: String,
+)
```
Would it worth using  `net.corda.libs.packaging.core.CpiIdentifier` instead?

vlajos commented on libs/ledger/ledger-common-impl/src/main/kotlin/net/corda/ledger/common/impl/transaction/CpkSummary.kt:
```diff
@@ -4,8 +4,9 @@ import net.corda.v5.base.annotations.CordaSerializable
+    val version: String,
+    val signerSummaryHash: String?,
+    val fileChecksum: String
+)
```
And `net.corda.libs.packaging.core.CpkIdentifier` instead of this?

vlajos commented on libs/ledger/ledger-consensual-impl/src/main/kotlin/net/corda/ledger/consensual/impl/transaction/ConsensualTransactionBuilderImpl.kt:
```diff
@@ -62,12 +74,46 @@ class ConsensualTransactionBuilderImpl(
+            name = cpiIdentifier.name,
+            version = cpiIdentifier.version,
+            signerSummaryHash = cpiIdentifier.signerSummaryHash?.toHexString(),
+            fileChecksum = "00000000000000000000000000000000FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF"
```
Do we need the cpi's file checksum?

omfR3 replied:
Good question -- one for @mnesbit I guess.

omfR3 committed:
Merge branch 'release/os/5.0' of github.com:corda/corda-runtime-os into CORE-5940/encode-cpk-metadata-in-consensual-transactions

yift-r3 committed:
CORE-7041: Remove warning from startup

omfR3 committed:
Merge branch 'yift/core-7041/remove-warning' of github.com:corda/corda-runtime-os into CORE-5940/encode-cpk-metadata-in-consensual-transactions

omfR3 committed:
CORE-5094: (ledger-consensual-impl) Use List rather than abstract Collection type to get around Kryoserializer error

omfR3 committed:
CORE-5940: (ledger-consensual) Convert Collection<T> -> List<T> more neatly

lankydan commented on libs/ledger/ledger-common-impl/src/main/kotlin/net/corda/ledger/common/impl/transaction/TransactionMetaData.kt:
```diff
@@ -28,16 +29,63 @@ class TransactionMetaData(
+                    when (it) {
+                        is CpkSummary -> it
+                        is Map<*, *> -> {
+                            CpkSummary(
+                                it["name"].toString(),
+                                it["version"].toString(),
+                                it["signerSummaryHash"]?.toString(),
+                                it["fileChecksum"].toString()
```
What happens if these don't exist?

omfR3 replied:
Then the data was incorrectly serialised and this will throw an NPE. Maybe we can wrap it and throw a `CordaRuntimeException` like in `getCpiMetadata` to make it less confusing. This is why I don't really like manually deserialising these weakly-typed maps -- it'd be better to have the JSON deserialiser validate the fields etc. We have a separate task to look at that.

lankydan replied:
ok

omfR3 replied:
Good point though. I've updated this and got rid of the super-weakly-typed `.toString()` conversions (since e.g. a missing `version` field would actually turn into a version string `null`, which is awful). Now it asserts the existence and stringiness of each (mandatory) field.

omfR3 committed:
Merge branch 'release/os/5.0' of github.com:corda/corda-runtime-os into CORE-5940/encode-cpk-metadata-in-consensual-transactions

omfR3 committed:
CORE-5940: (ledger-common) Better CPI/CPK info validation in TX metadata

When converting TX metadata, don't accept e.g. a null version and turn it into the String "null". Check all fields and throw sensible runtime error if any are bad (missing or wrong type).

omfR3 committed:
Merge branch 'release/os/5.0' of github.com:corda/corda-runtime-os into CORE-5940/encode-cpk-metadata-in-consensual-transactions

blsemo commented on components/virtual-node/ledger-persistence-service-impl/src/main/kotlin/net/corda/ledger/persistence/impl/internal/ConsensualLedgerRepository.kt:
```diff
@@ -154,7 +153,7 @@ class ConsensualLedgerRepository(
         tx: WireTransaction
     ) {
         // TODO get values from transaction metadata
```
Remove TODO comment?

omfR3 replied:
There is a bit of work that'll need to be done here to actually use those values. I think @mbrkic-r3 has probably already done it in his branch, though, so this comment can go.

blsemo commented on libs/ledger/ledger-common-impl/src/main/kotlin/net/corda/ledger/common/impl/transaction/CpkSummary.kt:
```diff
@@ -4,8 +4,9 @@ import net.corda.v5.base.annotations.CordaSerializable
-    val signerSummaryHash : String?
-)
\ No newline at end of file
+data class CpkSummary(
```
Contentwise, this looks exactly the same as a `CpiSummary`. Could we just have one (e.g. `CordaPackageSummary`) and use it in two places?

omfR3 replied:
Yes, fair point. Initially `CpiSummary` had an additional property `cpks: List<CpkSummary>`, but I've removed that and flattened them out. Will merge them together then.

blsemo commented on libs/ledger/ledger-common-impl/src/main/kotlin/net/corda/ledger/common/impl/transaction/TransactionMetaData.kt:
```diff
@@ -28,15 +30,66 @@ class TransactionMetaData(
+
+    fun getLedgerVersion(): String = this[LEDGER_VERSION_KEY].toString()
+
+    fun getCpiMetadata(): CpiSummary? {
```
This code seems a bit uncomfortable. Can we not define what format the CPI and CPK metadata will be packaged in? Or use a layered property map?
This at least needs a comment why the data could be in different formats.

omfR3 replied:
What do you mean by different formats -- like the fact that we might get back something that's already a `CpiSummary`, but it might also be a `Map<String, Any?>`? From what I can tell, this is a side effect of how we're doing JSON serialisation of the transaction metadata which is typed as `Map<String, Any?>`, and Jackson doesn't know in advance what the `Any?` is, causing it to be returned by the object mapper as an untyped map.

I would much prefer to define the format as a normal data class, but it seems we'd already achieved some consensus to store everything in a map (to reduce the risk of breaking user code when we add new properties to `TransactionMetaData`, if I understand correctly).

I'm not sure what a layered property map is; will investigate that as a possible solution. Also I notice now that we can refactor out some of this code now that `CpiSummary` and `CpkSummary` are the same and can be unified.

omfR3 replied:
@blsemo I've refactored some of that out. Can we leave further improvements for a separate PR later? This has been open for a while and a few other PRs are blocked waiting on it.

omfR3 committed:
CORE-5940: (ledger-common + consensual) PR feedback improvements

* Fuse CP(I|K)Summary into CordaPackageSummary now that they're identical.
* Refactor common construction code for CordaPackageSummary into static
  factory method "from(...)".

blsemo commented on libs/ledger/ledger-consensual-impl/src/main/kotlin/net/corda/ledger/consensual/impl/transaction/factory/ConsensualTransactionBuilderFactoryImpl.kt:
```diff
@@ -28,17 +28,10 @@ class ConsensualTransactionBuilderFactoryImpl @Activate constructor(
-            .filter{ it.isContractCpk() }
-            .map{ it.cpkId }
-            .toList()
+        val sandboxGroupContext = flowFiberService.getExecutingFiber().getExecutionContext().sandboxGroupContext
```
Probably ok for now - needs a refactor once Dan's SandboxGroupContextService is available.
